packages/internal/src/generate/templates/all-currentUser.d.ts.template

```diff
@@ -6,9 +6,20 @@ import { getCurrentUser } from '../../../api/src/lib/auth'
 type ThenArg<T> = T extends PromiseLike<infer U> ? U : T
 export type InferredCurrentUser = ThenArg<ReturnType<typeof getCurrentUser>>
 
+type UndefinedRoles = {
+  /**
+   * ⚠️ Heads-Up: This is undefined unless roles key is returned from {@link getCurrentUser()}
+   *
+   * You may look into to https://redwoodjs.com/docs/tutorial/chapter7/rbac for a tutorial on
+   * how to add fully-fledged RBAC (Role-based Access Control) to your database model.
+   */
+  roles?
```

Comment:
This has an implicit `any` type. Can we be explicit about the type? Otherwise @dac09 is just going to have to fix it in his Strict Mode PR

Reply:
I think this is actually a good thing. No need to type something we don't know the type of

Reply:
Is there a difference between an implicit and explicit `any`?
I thought the behaviour would be the same, only with the explicit `any` we're saying "Yes, I know this will be typed as `any`", and we make it more clear for all the readers of the code, by just glancing, that it's `any`.

But you're better with TS than I am Danny, so please tell me if I'm wrong. Eager to learn from you 🙂

Reply:
This syntax was new to me too! I thought it was pretty cool, personally. I think using `any` will also throw a linting error, but not super sure. 

The other alternative is to type it as `string | string[]`, or even `unknown`.

🤷 i'm not super opinionated either way on this, the key is that the actual return type from getCurrentUser will overwrite it with the real type.

Reply:
Also noone is better at TS, TS is better than you.

Reply:
I played around with this code in the [TS playground](https://www.typescriptlang.org/play) yesterday and actually made it `string | string[]` when I was experimenting. But didn't feel confident enough to propose that we use that. Thanks for validating my ideas Danny 🙂 But now that you mentioned `unknown` I think that's the better choice.

Reply:
`unknown` is a can do – (other) stuff turns red then, which brings us back to square 0.5:

![grafik](https://user-images.githubusercontent.com/1634615/177218152-0341c81e-9973-43fb-8373-3098a4900c25.png)

> Also noone is better at TS, "TS is better than <TYou>".

Whenever i think i got something right, TS comes around the corner again and smacks me good.  

Only Chuck Norris is better at TS.

Reply:
> I think this is actually a good thing.

Me too b/c it's the closest to "nothing" (what it really is until you set up RBAC, respectively change the result of `getCurrentUser`. Before that, it will never be `string | string[]`.

>  I think using any will also throw a linting error, but not super sure.

What about leaving it as it is for now and dealing with it on demand, meaning when we are sure this is needed during the cause of #5491 ?

Reply:
> Me too b/c it's the closest to "nothing"

For that we have `never`, but that won't fly. It'll give you at least as many red squigglies as `unknown` 🙁 

> What about leaving it as it is for now and dealing with it on demand, meaning when we are sure this is needed during the cause of https://github.com/redwoodjs/redwood/pull/5491 ?

Sure 🙂

Reply:
Hold on a sec.... just discovered that `unknown[]` would also work – funny but makes sense, b/c it could be an array of characters (so you can do string operations on it), as well as an array of anything else. I guess i'll just use that.

> I think using any will also throw a linting error, but not super sure.

@dac09 I just did quick research – it won't: https://dev.to/jsdev/strict-mode-typescript-j8p#1-noimplicitany
Wicked – turning on strict mode actually turns things upside down, implicit becomes explicit. We can't use explicit `any` in the project now, b/c that triggers the [no-explicit-any](https://typescript-eslint.io/rules/no-explicit-any/) es-lint rule. Once we'd go strict in #5491, what is currently forbidden becomes mandatory :exploding_head:

Reply:
Closing this conversation via b551859a43126de4b0bd301432cd9243e6a11dc6

Reply:
So should the type perhaps be
’unknown | unknown[]’ ?

Reply:
> So should the type perhaps be `unknown | unknown[]`

Been there, done that, boom: all goes red again – it seems that in this case `unknown` takes precedence, no matter how i order the two.

#theSubletiesOfTypeScript :shrug:

Reply:
> What version of RW have you been testing your code on?

Oh dang, stupid me ... that's from my bread-and-butter project that was created on v1.0.0 :facepalm: 
With the recent code, `unknown[]` will still throw a type error. Excellent catch. 

> And I think that'd work with unknown | unknown[]

It does! As said [above](https://github.com/redwoodjs/redwood/pull/5856#discussion_r913288177), it behaves *exactly* the same as if i just wrote (even just shows `unknown` on the code completion tooltip for `roles`). However, for the sake of clarity and consistency in case anybody clicks on it and navigates to `all-currentUser.d.ts`, i'm adding it as a union type now.